### PR TITLE
Load balancing manual priority

### LINKED ISSUE #11565: Extend load_balancing first_or_random to first_2th_or_random
**Use case**
In https://clickhouse.tech/docs/en/operations/settings/settings/#load_balancing-first_or_random, `first_or_random` works for 1 replica per AZ, but doesn't work for 2 replicas per AZ.
Let's say, there are two AZs (A and B), and 1 shard and 2 replicas in each AZ. 4 hosts are like below.
```
A_shard1_replicas1
A_shard1_replicas2
B_shard1_replicas1
B_shard1_replicas2
```
Configure load_balancing = first_or_random
In AZ A, remote_servers.xml is 
```
<yandex>
    <remote_servers>
        <cross-az>
            <shard>
                <internal_replication>true</internal_replication>
                <replica>
                    <host>A_shard1_replicas1</host>
                    <port>9000</port>
                </replica>
                <replica>
                    <host>A_shard1_replicas2</host>
                    <port>9000</port>
                </replica>
                <replica>
                    <host>B_shard1_replicas1</host>
                    <port>9000</port>
                </replica>
                <replica>
                    <host>B_shard1_replicas2</host>
                    <port>9000</port>
                </replica>
            </shard>
        </cross-az>
    </remote_servers>
</yandex>
```
In AZ B, remote_servers.xml is
```
<yandex>
    <remote_servers>
        <cross-az>
            <shard>
                <internal_replication>true</internal_replication>
                <replica>
                    <host>B_shard1_replicas1</host>
                    <port>9000</port>
                </replica>
                <replica>
                    <host>B_shard1_replicas2</host>
                    <port>9000</port>
                </replica>
                <replica>
                    <host>A_shard1_replicas1</host>
                    <port>9000</port>
                </replica>
                <replica>
                    <host>A_shard1_replicas2</host>
                    <port>9000</port>
                </replica>
            </shard>
        </cross-az>
    </remote_servers>
</yandex>
```
In AZ A, if `A_shard1_replicas1` is unavailable, then the first_or_random algorithm chooses randomly from the left 3 hosts, but it is better to choose `A_shard1_replicas2`. 

**Describe the solution you'd like**
In AZ A, we want first_2th_or_random load_balance, which will act as below:
- If `A_shard1_replicas1` and `A_shard1_replicas2` are available, randomly choose one of them, 
- If one of `A_shard1_replicas1` and `A_shard1_replicas2` is available, choose the available one,
- If `A_shard1_replicas1` and `A_shard1_replicas2` are unavailable, randomly choose one of `B_shard1_replicas1` and `B_shard1_replicas2`

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add replica priority for load_balancing (for manual prioritization of the load balancing).

Detailed description / Documentation draft:
Can be used for:
- excluding some hosts until there are no errors for other replicas (or you can ignore some errors with `distributed_replica_max_ignored_errors`)

Fixes: #11565

### CODE REVIEW

**Comment by azat** [comment]
> test_adaptive_granularity/test.py::test_version_update_two_nodes

Is in ci fest

**Comment by alexey-milovidov** [comment]
> Integration tests

Broken in #11903

**Comment by azat** [comment]
Will rebase against uptream/master, since I moved one patch(e02cf7a652eba0e6e72e4dcd92d381727ab9bd12)  into #12040 (did not want to create separate PR just for this, but two unrelated changes is worse)

File: docs/en/engines/table-engines/special/distributed.md [hunk 1]
```diff
@@ -50,6 +50,8 @@ Clusters are set like this:
             <!-- Optional. Whether to write data to just one of the replicas. Default: false (write data to all replicas). -->
             <internal_replication>false</internal_replication>
             <replica>
+                <!-- Optional. Priority of the replica for load balancing (see also load_balancing setting). Default: 1 (less value has more priority). -->
+                <priority>1</priority>
```
Comment (filimonov) [inline docs/en/engines/table-engines/special/distributed.md:54]: Always that silly problem with what has bigger priority 0 or 255.

If you prefer to have less value = bigger priority, please ensure that we support here negative numbers. (user can want to increase the priority "above" 1 or zero).

Comment (azat) [reply docs/en/engines/table-engines/special/distributed.md:54]: >ensure that we support here negative numbers

There is simple sort, should work

**Comment by azat** [comment]
Any news on this one? (This will conflicts with #11498 )

**Comment by azat** [comment]
@nikitamikhaylov friendly ping

**Comment by nikitamikhaylov** [comment]
> @nikitamikhaylov friendly ping

😉

**Review by nikitamikhaylov — APPROVED** [review]

**Pull request merged**